### PR TITLE
Feature/update php cs fixer 2.8.x

### PR DESCRIPTION
**Pull request by toofff**
The project [php-cs-fixer-styleci-bridge](https://github.com/Soullivaneuh/php-cs-fixer-styleci-bridge) has not been released for several months.
When you run php-cs-fix in version 2, nothing works.

I propose to you to pass this project directly and the others with a configuration for php-cs-fix version 2.

On the other hand for the configuration I am maybe going a little strong?

Closes https://github.com/sonata-project/dev-kit/issues/158
Closes https://github.com/sonata-project/dev-kit/issues/157

### CODE REVIEW

File: .php_cs.dist [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        'ordered_imports' => true,
+        'phpdoc_order' => true,
+        'psr4' => true,
+        'semicolon_after_instruction' => true,
+        'strict_comparison' => true,
+        'strict_param' => true,
+        'ternary_to_null_coalescing' => true,
+    ])
+    ->setUsingCache(true)
+;
```
Comment (greg0ire) [inline .php_cs.dist]: Please configure your IDE to always add a newline at end of file, this will make the red pictogram disappear.

Comment (toofff) [reply .php_cs.dist]: Ooops I forgot to do it for this project.

Comment (toofff) [reply .php_cs.dist]: Done

**Comment by core23** [comment]
So we must remove StyleCI integration after merging this?

File: project/.php_cs.dist [hunk 2]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+/*
+ * DO NOT EDIT THIS FILE!
+ *
+ * It's auto-generated by sonata-project/dev-kit package.
+ *
+ * Package `sllh/php-cs-fixer-styleci-bridge` is required to get it working.
```
Comment (core23) [inline project/.php_cs.dist]: This is still true?

File: .php_cs.dist [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    ->in(__DIR__)
+    ->exclude('Tests/Fixtures')
+    ->exclude('tests/Fixtures')
+    ->exclude('Resources/skeleton')
+    ->exclude('Resources/public/vendor')
+;
+
+return PhpCsFixer\Config::create()
+    ->setFinder($finder)
+    ->setRiskyAllowed(true)
```
Comment (greg0ire) [inline .php_cs.dist]: I'm not sure this should be true... will it create BC-breaks by e.g. changing signatures to add `void`, things like that?

File: .php_cs.dist [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    ->setRules([
+        '@Symfony' => true,
+        '@Symfony:risky' => true,
```
Comment (greg0ire) [inline .php_cs.dist]: What does this do exactly?

File: project/.php_cs.dist (hunk 2 above)
Comment (toofff) [reply project/.php_cs.dist]: It's just a bad copy paste from the old file.

File: .php_cs.dist (hunk 3 above)
Comment (toofff) [reply .php_cs.dist]: I set the `setRiskyAllowed` option to `true` to use the options below :
- declare_strict_types
- modernize_types_casting
- psr4
- strict_comparison
- strict_param

But is this really necessary for the different Sonata project?

**Comment by toofff** [comment]
Yes it will remove StyleCI integration after merging.
I forgot to do it yesterday, but can I add PHPCSFIxer in travis?
What do you think @core23 @greg0ire ?

**Comment by Soullivaneuh** [comment]
> The project php-cs-fixer-styleci-bridge has not been released for several months.

Not released from month does not mean this one is not maintained at all. Otherwise, I would put a message about it. :wink: 

> When you run php-cs-fix in version 2, nothing works.

Well, simply because the StyleCI configuration has diverged form php-cs-fixer v2. :confused: 

https://github.com/Soullivaneuh/php-cs-fixer-styleci-bridge/pull/79

And AFAIK, nothing changed on this side.

I personnaly :-1: to remove it right know and would like to change the CI instead (I'm working on a new one) to still have one working waiting that.

But if the @sonata-project/contributors vote to remove it right know, I'm ok with it.

File: .php_cs.dist (hunk 3 above)
Comment (Soullivaneuh) [reply .php_cs.dist]: The `setRiskyAllowed` can stay here, but we have to filter the risky fixers.

For example, `strict_param` and `strict_comparison` might introduce BC break on current stables.

File: .php_cs.dist [hunk 5]
```diff
@@ -0,0 +1,60 @@
+<?php
```
Comment (Soullivaneuh) [inline .php_cs.dist]: Please make a link to `project/.php_cs.dist` or require it we should have the same configuration.

File: .php_cs.dist [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    ->setFinder($finder)
+    ->setRiskyAllowed(true)
+    ->setRules([
+        '@Symfony' => true,
+        '@Symfony:risky' => true,
+        'array_syntax' => [
+            'syntax' => 'short',
+        ],
+        'braces' => [
+            'allow_single_line_closure' => true,
```
Comment (Soullivaneuh) [inline .php_cs.dist]: Is it compliant to SF standards?

File: .php_cs.dist [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        ],
+        'declare_strict_types' => true,
+        'header_comment' => [
+            'header' => $header,
+            'location' => 'after_open',
+        ],
+        'modernize_types_casting' => true,
+        'no_extra_consecutive_blank_lines' => [
+            'break',
```
Comment (Soullivaneuh) [inline .php_cs.dist]: Why this list? Can't we just enable the fixer globally?

File: .php_cs.dist [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+            'continue',
+            'curly_brace_block',
+            'extra',
+            'parenthesis_brace_block',
+            'return',
+            'square_brace_block',
+            'throw',
+            'use',
+        ],
+        'no_useless_else' => true,
```
Comment (Soullivaneuh) [inline .php_cs.dist]: AFAIK, this one is buggy: https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/2694

File: .php_cs.dist [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        'no_useless_return' => true,
+        'ordered_imports' => true,
+        'phpdoc_order' => true,
+        'psr4' => true,
+        'semicolon_after_instruction' => true,
+        'strict_comparison' => true,
```
Comment (Soullivaneuh) [inline .php_cs.dist]: Please remove it. This one may introduce BC breaks.

File: .php_cs.dist [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        'strict_param' => true,
```
Comment (Soullivaneuh) [inline .php_cs.dist]: Please remove it. This one may introduce BC breaks.

File: .php_cs.dist [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+            'square_brace_block',
+            'throw',
+            'use',
+        ],
+        'no_useless_else' => true,
+        'no_useless_return' => true,
+        'ordered_imports' => true,
+        'phpdoc_order' => true,
+        'psr4' => true,
+        'semicolon_after_instruction' => true,
```
Comment (Soullivaneuh) [inline .php_cs.dist]: This one is already included on `Symfony` rules set, please remove.

File: .php_cs.dist [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+            'return',
+            'square_brace_block',
+            'throw',
+            'use',
+        ],
+        'no_useless_else' => true,
+        'no_useless_return' => true,
+        'ordered_imports' => true,
+        'phpdoc_order' => true,
+        'psr4' => true,
```
Comment (Soullivaneuh) [inline .php_cs.dist]: To be remove if we keep `Symfony:risky`.

File: web/index.php [hunk 13]
```diff
@@ -9,6 +9,8 @@
  * file that was distributed with this source code.
  */
 
+declare(strict_types=1);
```
Comment (Soullivaneuh) [inline web/index.php]: Remove this `declare` from dev-kit files. It should not be introduced with common cs files.

**Comment by Soullivaneuh** [comment]
As PHP 7.1 will be required on next major, we may take benefit to introduce really risky rules when PHP version is 7.1+?

File: src/Config/DevKitConfiguration.php [hunk 14]
```diff
@@ -42,8 +42,7 @@ public function getConfigTreeBuilder()
                 ->arrayNode('packages')
                     ->prototype('scalar')->end()
                 ->end()
-            ->end()
-        ;
```
Comment (Soullivaneuh) [inline src/Config/DevKitConfiguration.php]: Which rule is doing that? It should not be changed.

File: src/Console/Command/DispatchCommand.php [hunk 15]
```diff
@@ -124,11 +122,11 @@ private function updateRepositories(Package $package, array $projectConfig)
 
         $repositoryInfo = $this->githubClient->repo()->show(static::GITHUB_GROUP, $repositoryName);
         $infoToUpdate = [
-            'homepage' => 'https://sonata-project.org/',
-            'has_issues' => true,
-            'has_projects' => true,
-            'has_wiki' => false,
-            'default_branch' => end($branches),
+            'homepage'           => 'https://sonata-project.org/',
```
Comment (Soullivaneuh) [inline src/Console/Command/DispatchCommand.php]: Array must not be aligned.

File: composer.json [hunk 16]
```diff
@@ -17,10 +17,10 @@
         "php-http/guzzle6-adapter": "^1.1",
         "sebastian/diff": "^1.4",
         "silex/silex": "^2.0",
-        "sllh/php-cs-fixer-styleci-bridge": "^2.0",
         "symfony/config": "^3.0",
         "symfony/console": "^3.0",
         "symfony/filesystem": "^3.0",
+        "symfony/yaml": "^3.0",
```
Comment (Soullivaneuh) [inline composer.json]: Well, why not get the latest?

File: project/.php_cs.dist [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    ->setRules([
+        '@Symfony' => true,
+        'braces' => [
+            'allow_single_line_closure' => true,
+        ],
+        'header_comment' => [
+            'header' => $header,
+            'location' => 'after_open',
+        ],
+        'no_extra_consecutive_blank_lines' => [
```
Comment (Soullivaneuh) [inline project/.php_cs.dist]: Please leave this at `true` to keep the Symfony default config.

File: project/.php_cs.dist [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    ->exclude('Tests/Fixtures')
+    ->exclude('tests/Fixtures')
+    ->exclude('Resources/skeleton')
+    ->exclude('Resources/public/vendor')
+;
+
+return PhpCsFixer\Config::create()
+    ->setFinder($finder)
+    ->setRules([
+        '@Symfony' => true,
```
Comment (Soullivaneuh) [inline project/.php_cs.dist]: Could you please add the following rules set?

- Symfony:risky
- PHP56Migration
- PHP56Migration:risky
- PHPUnit57Migration:risky

**Comment by toofff** [comment]
@Soullivaneuh I will take time at the lunch break to finalize the last comments.

**Comment by toofff** [comment]
@Soullivaneuh I think I have corrected all your requests.


By cons I do not feel that it works the symbolic link for the file `.php_cs.dist` at the root of the project. Because it should have returned an error with the current configuration.
The only solution I find is to duplicate the `.php_cs.dist` file at the root of the project.


Do I need to do other PRs in other Sonata projects to remove dependency `sllh/php-cs-fixer-styleci-bridge`, for example [the projet SonataAdminBundle](https://github.com/sonata-project/SonataAdminBundle/blob/3.x/composer.json#L50) ?

File: project/.travis.yml.twig [hunk 19]
```diff
@@ -75,6 +75,7 @@ matrix:
 
 before_install:
   - git remote add upstream ${UPSTREAM_URL} && git fetch --all
+  - if [[ $lint = 1 ]]; then wget https://github.com/FriendsOfPHP/PHP-CS-Fixer/releases/download/v2.8.1/php-cs-fixer.phar; fi
```
Comment (Soullivaneuh) [inline project/.travis.yml.twig]: Use http://cs.sensiolabs.org/download/php-cs-fixer-v2.phar instead.

File: .travis.yml [hunk 20]
```diff
@@ -14,10 +14,12 @@ cache:
 
 before_install:
   - phpenv config-rm xdebug.ini
+  - wget http://cs.sensiolabs.org/download/php-cs-fixer-v2.phar
```
Comment (Soullivaneuh) [inline .travis.yml]: `--output-document php-cs-fixer.phar`

File: .php_cs.dist (hunk 5 above)
Comment (Soullivaneuh) [reply .php_cs.dist]: Please use require to the project configuration.

**Comment by toofff** [comment]
@Soullivaneuh If you do not mind, I'd really like to finish it myself.

For your [comment #336](https://github.com/sonata-project/dev-kit/pull/336#issuecomment-343656138), if I understand correctly, you want PHP 7.1 to be required?

So I just have to delete the rules for PHP 5.x and add the rules for PHP 7.1 in the .php_cs.dist file? 
I also have to correct the errors that will be reported by PHP-Cs-Fixer after launching it with the new rules for PHP 7.1?
If it's good for you, I'm doing it now.

**Comment by Soullivaneuh** [comment]
> @Soullivaneuh If you do not mind, I'd really like to finish it myself.

No problem, just help proposing. :+1: 

> For your comment #336, if I understand correctly, you want PHP 7.1 to be required?

No. I mean, update the configuration to add the risky rules en PHP migration rules if php-cs-fixer is run under PHP 7.1.

> I also have to correct the errors that will be reported by PHP-Cs-Fixer after launching it with the new rules for PHP 7.1?

I don't know, it could be fixed on dev-kit PR. @sonata-project/contributors WDYT?

In any case, please wait this PR being merged before doing anything.

> Do I need to do other PRs in other Sonata projects to remove dependency sllh/php-cs-fixer-styleci-bridge, for example the projet SonataAdminBundle ?

If you want to do it, you can. But after getting this PR merged and deployed across repositories.

**Comment by toofff** [comment]
@Soullivaneuh 

I found the [problem with the symbolic link](https://github.com/sonata-project/dev-kit/pull/336/files#diff-ce61b3e08dc11c4c455ab5dcbaa300c6R18), it was a simple problem of path.

File: .travis.yml [hunk 21]
```diff
@@ -14,10 +14,12 @@ cache:
 
 before_install:
   - phpenv config-rm xdebug.ini
+  - wget http://cs.sensiolabs.org/download/php-cs-fixer-v2.phar --output-document php-cs-fixer.phar
```
Comment (dunglas) [inline .travis.yml]: I recommend to download the PHAR from GitHub (https://github.com/FriendsOfPHP/PHP-CS-Fixer/releases) because:

* It is available trough HTTPS
* You can stick to a specific version, it makes the tests more reliable

File: project/.travis.yml.twig [hunk 22]
```diff
@@ -75,6 +75,7 @@ matrix:
 
 before_install:
   - git remote add upstream ${UPSTREAM_URL} && git fetch --all
+  - if [[ $lint = 1 ]]; then wget http://cs.sensiolabs.org/download/php-cs-fixer-v2.phar --output-document php-cs-fixer.phar; fi
```
Comment (dunglas) [inline project/.travis.yml.twig]: Maybe am I missing something, but it looks like the `$lint` var is always undefined.

File: src/Console/Command/DependsCommand.php [hunk 23]
```diff
@@ -37,7 +37,7 @@ protected function configure()
      */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
-        $branchDepth = intval($input->getOption('branch-depth'));
+        $branchDepth = (int) ($input->getOption('branch-depth'));
```
Comment (dunglas) [inline src/Console/Command/DependsCommand.php]: Parenthesis around `$input->getOption('branch-depth')` are useless.

**Comment by dunglas** [comment]
For what it's worth, API Platform switched from StyleCI to PHP CS Fixer some times ago too, because PHP CS Fixer is more flexible (especially when using PHP 7+).

**Comment by Soullivaneuh** [comment]
I agree @dunglas, it's quite bad than StyleCI diverged from php-cs-fixer v2. It's also why the bridge I written is hard to maintain.

I'm currently trying a work to make a similar tool but running php-cs-fixer (and others) natively.

Thanks for the review BTW! :+1:

File: .travis.yml [hunk 24]
```diff
@@ -14,10 +14,12 @@ cache:
 
 before_install:
   - phpenv config-rm xdebug.ini
+  - wget https://github.com/FriendsOfPHP/PHP-CS-Fixer/releases/download/v2.8.2/php-cs-fixer.phar
```
Comment (Soullivaneuh) [inline .travis.yml]: I would like to just download `v2`, minor version should be BC.

File: composer.json [hunk 25]
```diff
@@ -10,22 +10,22 @@
...
-        "silex/silex": "^2.0",
-        "sllh/php-cs-fixer-styleci-bridge": "^2.0",
-        "symfony/config": "^3.0",
-        "symfony/console": "^3.0",
-        "symfony/filesystem": "^3.0",
-        "twig/twig": "^1.24",
-        "vlucas/phpdotenv": "^2.2"
+        "sebastian/diff": "^2.0",
+        "silex/silex": "^2.2",
+        "symfony/config": "^3.3",
```
Comment (Soullivaneuh) [inline composer.json]: Why those changes? They are not part of the PR, please revert and just keep `php-cs-fixer-styleci-bridge` removal.

File: project/.php_cs.dist [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+        'location' => 'after_open',
+    ],
+    'no_extra_consecutive_blank_lines' => true,
+    'no_useless_return' => true,
+    'ordered_imports' => true,
+    'phpdoc_order' => true,
+    'ternary_to_null_coalescing' => true,
+];
+
+if (\PHP_VERSION_ID >= 50600 && \PHP_VERSION_ID < 70000) {
```
Comment (Soullivaneuh) [inline project/.php_cs.dist]: AFAIK, we can use the migration below even with PHP7. Please remove `\PHP_VERSION_ID < 70000`

File: project/.php_cs.dist [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    $rules = array_merge($rules, [
+        '@PHP56Migration' => true,
+        '@PHP56Migration:risky' => true,
+        '@PHPUnit57Migration:risky' => true,
+    ]);
+} else if (\PHP_VERSION_ID >= 70000 && \PHP_VERSION_ID < 70100) {
```
Comment (Soullivaneuh) [inline project/.php_cs.dist]: Remove `\PHP_VERSION_ID < 70100`

And simple if can be used.

File: project/.php_cs.dist [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+    $rules = array_merge($rules, [
+        'PHP70Migration:risky' => true,
+    ]);
+} else if (\PHP_VERSION_ID >= 70100 && \PHP_VERSION_ID < 70200) {
```
Comment (Soullivaneuh) [inline project/.php_cs.dist]: Remove `PHP_VERSION_ID < 70200` and use simple `if`.

File: project/.travis.yml.twig [hunk 29]
```diff
@@ -49,6 +49,7 @@ matrix:
 {% if docs_target %}
     - php: '{{ php|last }}'
       env: TARGET=lint
+      env: lint=1
```
Comment (Soullivaneuh) [inline project/.travis.yml.twig]: We don't need this variable.

You should instead download php-cs-fixer binary on this file: https://github.com/sonata-project/dev-kit/blob/master/project/.travis/install_lint.sh (you may use the already present `composer global require`).

And add the php-cs-fixer command here: https://github.com/sonata-project/dev-kit/blob/5d31b22460395a1bc65c077f9dee1e122cc70f09/project/Makefile#L16 (BTW, you don't need the `--dry-run and the --diff` as we already check the diff at the end.

Sorry for that, I forgot to tell you.

File: project/.php_cs.dist [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+];
+
+if (\PHP_VERSION_ID >= 50600 && \PHP_VERSION_ID < 70000) {
+    $rules = array_merge($rules, [
+        '@PHP56Migration' => true,
+        '@PHP56Migration:risky' => true,
+        '@PHPUnit57Migration:risky' => true,
+    ]);
+} else if (\PHP_VERSION_ID >= 70000 && \PHP_VERSION_ID < 70100) {
+    $rules = array_merge($rules, [
```
Comment (Soullivaneuh) [inline project/.php_cs.dist]: strict type declare must be added.

**Comment by Soullivaneuh** [comment]
I'll continue and finish your PR in #358, we need it for user-bundle.

Thanks a lot for your work! :+1:

**Comment by toofff** [comment]
@Soullivaneuh 

I did not have time to look at it this weekend.
I wanted to propose to you to finish it, so that it is easier for you.

Thanks and see you soon on an upcoming pull request ;)
